myia/abstract/utils.py

```diff
@@ -195,20 +197,169 @@ def build_type_fn(self, x: Function):
     return dtype.Function
 
 
+############
+# Checking #
+############
+
+
+@dataclass
+class CheckState:
+    """State of abstract_check."""
+
+    cache: dict
+    prop: str
+
+
+@overload.wrapper(initial_state=lambda: None)
+def abstract_check(__call__, self, x, *args):
+    """Check that the given object has a certain property."""
+    def proceed():
+        if hasattr(x, prop):
+            return getattr(x, prop) is x
+        elif __call__(self, x, *args):
+            if isinstance(x, (AbstractValue, Context, Reference)):
+                setattr(x, prop, x)
```

Comment:
This also feels specific to `is_concrete`.

Reply:
Yeah, maybe, but I'm not sure what other criterion to put. I guess try/catch would work, but some classes might raise some exotic exception here.